Bulldozer.CSV/CSVComponent.cs

```diff
@@ -488,804 +836,2680 @@ public static int GetAttributeFieldType( string attributeTypeString )
-        private const int BenevolenceRequestZip = 21;                /* String [Optional] */
-        private const int BenevolenceRequestCountry = 22;            /* String [Optional] */
-        private const int BenevolenceType = 23;                      /* String [Optional: Rock v13+] */
+        /// <summary>
+        /// Loads the defined values.
+        /// </summary>
+        /// <param name="definedTypeGuid">The defined type unique identifier.</param>
+        /// <returns></returns>
+        private Dictionary<Guid, DefinedValueCache> LoadDefinedValues( Guid definedTypeGuid )
+        {
+            return DefinedTypeCache.Get( definedTypeGuid ).DefinedValues.ToDictionary( k => k.Guid );
+        }
+    }
 
-        #endregion BenevolenceRequest Constants
+    /// <summary>
+    /// Dictionary Extensions Helper Class
+    /// </summary>
+    public static class DictionaryExtensions
+    {
+        /// <summary>
+        /// Converts a DefinedValue dictionary (indexed by Guid) into a dictionary indexed by unique values.
+        /// </summary>
+        /// <param name="inputDictionary">The source dictionary (indexed by Guid).</param>
+        /// <returns></returns>
+        public static Dictionary<string, DefinedValueCache> GetUniqueValues( this Dictionary<Guid, DefinedValueCache> inputDictionary )
+        {
+            return inputDictionary.Values
+                .GroupBy( k => k.Value ).Select( grp => grp.First() )
+                .ToDictionary( v => v.Value, p => p, StringComparer.OrdinalIgnoreCase );
+        }
+    }
 
-        #region BenevolenceResult Constants
+    public class AttributeObject
+    {
+        public string AttributeId { get; set; }
 
-        /*
-         * This is the definition of the csv format for the BenevolenceResult.csv file.
-         *
-         */
+        public string AttributeName { get; set; }
 
-        private const int BenevolenceResultRequestId = 0;          /* String [Required] */
-        private const int BenevolenceResultType = 1;               /* String | Int [Required] */
-        private const int BenevolenceResultId = 2;                 /* String [Optional] */
-        private const int BenevolenceResultAmount = 3;             /* Decimal [Optional] */
-        private const int BenevolenceResultSummary = 4;            /* String [Optional] */
-        private const int BenevolenceResultCreatedById = 5;        /* String [Optional] */
-        private const int BenevolenceResultCreatedDate = 6;        /* DateTime [Optional] */
+        public string DefinedTypeId { get; set; }
+    }
 
-        #endregion BenevolenceResult Constants
+    public class AttributeValueObject
+    {
+        public string AttributeKey { get; set; }
 
-        #region Person History
+        public string AttributeValue { get; set; }
 
-        /*
-         * This is the definition of the csv format for the PersonHistory.csv file.
-         */
+        public int EntityTypeId { get; set; }
```

Comment:
@nateh777  Yes. Since Rock does not prohibit duplicate attribute keys, and since we cannot guarantee duplicate keys are not coming in through our data, and since we potentially have attributes/values coming in from multiply entity types, we are trying to make our attribute associations as strong as possible. This property is used in the AddAttributeValueDVs method to build a complex key for a complicated dictionary linking attribute values to their defined type values. Attribute keys are combined with entitytypeid to provide as much distinction as possible.